vms/platformvm/validators/fee/fee.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fee
+
+import (
+	"math"
+
+	"github.com/ava-labs/avalanchego/vms/components/gas"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+)
+
+// CalculateExcess updates the excess value based on the target and current
+// usage over the provided duration.
+func CalculateExcess(
+	target gas.Gas,
+	current gas.Gas,
+	excess gas.Gas,
+	duration uint64,
+) gas.Gas {
+	if current < target {
+		return excess.SubPerSecond(target-current, duration)
+	}
+	return excess.AddPerSecond(current-target, duration)
+}
+
+// CalculateCost calculates how much to charge based on the dynamic fee
+// mechanism for [duration].
+func CalculateCost(
+	target gas.Gas,
+	current gas.Gas,
+	minPrice gas.Price,
+	excessConversionConstant gas.Gas,
+	excess gas.Gas,
+	duration uint64,
+) uint64 {
+	// If the current and target are the same, the price is constant.
+	if current == target {
+		price := gas.CalculatePrice(minPrice, excess, excessConversionConstant)
+		cost, err := safemath.Mul(duration, uint64(price))
+		if err != nil {
+			return math.MaxUint64
+		}
+		return cost
+	}
+
+	var (
+		cost uint64
+		err  error
+	)
+	for i := uint64(0); i < duration; i++ {
+		excess = CalculateExcess(target, current, excess, 1)
+		// If the excess is 0, the price will remain constant for all future
```

Comment:
This took a long time for me to grok (including asking you in a DM) so I think it's ok to be verbose in the explanation. Here's a first draft of my understanding that you can copy and tweak:

_The change in excess is positively correlated with the difference between current and target. It is also strictly bounded in [0,2^64) with (under/over)flows being clipped. As we know that current != target, excess would be strictly increasing or strictly decreasing if it weren't for clipping. However, when coupled with clipping, this guarantees that a zero excess will remain zero for all future iterations._